pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,444 @@
+		c.gwQueue.Forget(k)
+		klog.InfoS("Expected string in work queue but got", "object", k)
+		return true
+	} else if err := c.syncGateway(); err == nil {
+		c.gwQueue.Forget(key)
+	} else {
+		c.gwQueue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing Gateway, requeuing", "key", key)
+	}
+	return true
+}
+
+func (c *MCPodRouteController) syncGateway() error {
+	activeGW, err := getActiveGateway(c.gwLister, c.namespace)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get an active Gateway")
+		return err
+	}
+
+	c.podWorkersStartedMutex.Lock()
+	defer c.podWorkersStartedMutex.Unlock()
+	// Stop Pod flow controller and clean up all installed Multi-cluster Pod flows,
+	// if the Node was a Gateway before.
+	if activeGW == nil || c.nodeConfig.Name != activeGW.Name {
+		if c.podWorkersStarted {
+			klog.InfoS("Shutting down Multi-cluster PodFlowController")
+			close(c.podWorkerStopCh)
+			c.podInformer = nil
+			c.podLister = nil
+			c.podWorkersStarted = false
+		}
+		if len(c.podCache) > 0 {
+			err := c.ofClient.UninstallMulticlusterPodFlows("")
+			if err != nil {
+				return err
+			}
+			c.podCache = map[string]*podNodeIPs{}
+		}
+		return nil
+	}
+
+	if c.nodeConfig.Name == activeGW.Name {
+		if !c.podWorkersStarted {
+			klog.InfoS("Starting Multi-cluster PodFlowController")
+			c.podWorkerStopCh = make(chan struct{})
+			c.createPodInformer()
+			go c.podInformer.Run(c.podWorkerStopCh)
+			if !cache.WaitForNamedCacheSync(podRouteControllerName, c.podWorkerStopCh, c.podInformer.HasSynced) {
+				close(c.podWorkerStopCh)
+				c.podInformer = nil
+				c.podLister = nil
+				return errors.New("failed to sync Pod cache")
+			}
+
+			for i := 0; i < podWorkerNum; i++ {
+				go wait.Until(c.podWorker, time.Second, c.podWorkerStopCh)
+			}
+			c.podWorkersStarted = true
+			return nil
+		}
+		// Do nothing when the Pod flow controller is already started since
+		// Pod flow controller will be responsible for handling Pod events to install flows.
+	}
+	return nil
+}
+
+func (c *MCPodRouteController) cleanUpStalePodFlows() {
+	c.podWorkersStartedMutex.Lock()
+	defer c.podWorkersStartedMutex.Unlock()
+	if !c.podWorkersStarted && len(c.podCache) > 0 {
+		if err := c.ofClient.UninstallMulticlusterPodFlows(""); err == nil {
+			c.podCache = map[string]*podNodeIPs{}
+		}
+	}
+}
+
+func (c *MCPodRouteController) podWorker() {
+	for c.processPodNextWorkItem() {
+	}
+}
+
+func (c *MCPodRouteController) processPodNextWorkItem() bool {
+	obj, quit := c.podQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.podQueue.Done(obj)
+
+	if k, ok := obj.(*podNodeIPs); !ok {
+		c.podQueue.Forget(obj)
+		klog.InfoS("Expected *podNodeIPs in work queue but got", "object", obj)
+		return true
+	} else if err := c.syncPod(k); err == nil {
+		c.podQueue.Forget(k)
+	} else {
+		c.podQueue.AddRateLimited(k)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+	}
+	return true
+}
+
+func (c *MCPodRouteController) syncPod(ips *podNodeIPs) error {
```

Comment:
I feel we cannot use podNodeIPs as the key. The reason is the flow key (and OF match condition) is just Pod IP, so if the IP is reallocated, two workers can have a race condition to add/delete the same flow key, and even one worker can mistakenly override flow at failure retry.

Could you think about my earlier idea again:
> I remember earlier, you build the caches in enqueuePod(), and enqueue the Pod IP. What is wrong with that approach? In workers, you can just check the latest state of the Pod IP from caches, and decide what to do - delete or add flows for the IP.

We can discuss when you are available.

Reply:
yes, you  are right, actually only podIP is used in the syncPod, I forgot to update the key in the syncPod. I will update.

Reply:
Done.